pgjdbc/src/test/java/org/postgresql/core/AsciiStringInternerTest.java

```diff
@@ -0,0 +1,130 @@
+    String interned = interner.getString(bytes, 0, bytes.length, null);
+
+    //interned value should be equal
+    assertEquals(s1, interned);
+    //but should be different instance
+    assertNotSame(s1, interned);
+    //asking for it again, however should return same instance
+    assertSame(interned, interner.getString(bytes, 0, bytes.length, null));
+
+    //now show that we can get the value back from a different byte[]
+    byte[] bytes2 = new byte[128];
+    System.arraycopy(bytes, 0, bytes2, 73, bytes.length);
+    assertSame(interned, interner.getString(bytes2, 73, bytes.length, null));
+
+    //now we will mutate the original byte[] to show that does not affect the map
+    Arrays.fill(bytes, (byte) 13);
+    assertSame(interned, interner.getString(bytes2, 73, bytes.length, null));
+  }
+
+  @Test
+  public void testStagedValue() throws Exception {
+    AsciiStringInterner interner = new AsciiStringInterner();
+    String s1 = "testStagedValue";
+    interner.putString(s1);
+    byte[] bytes = s1.getBytes(StandardCharsets.US_ASCII);
+    String interned = interner.getString(bytes, 0, bytes.length, null);
+    // should be same instance
+    assertSame(s1, interned);
+    //asking for it again should also return same instance
+    assertSame(s1, interner.getString(bytes, 0, bytes.length, null));
+
+    //now show that we can get the value back from a different byte[]
+    byte[] bytes2 = new byte[128];
+    System.arraycopy(bytes, 0, bytes2, 73, bytes.length);
+    assertSame(s1, interner.getString(bytes2, 73, bytes.length, null));
+  }
+
+  @Test
+  public void testNonAsciiValue() throws Exception {
+    final Encoding encoding = Encoding.getJVMEncoding("UTF-8");
+    AsciiStringInterner interner = new AsciiStringInterner();
+    String s1 = "testNonAsciiValue" + '\u03C0'; // add multi-byte to string to make invalid for intern
+    byte[] bytes = s1.getBytes(StandardCharsets.UTF_8);
+    String interned = interner.getString(bytes, 0, bytes.length, encoding);
+
+    //interned value should be equal
+    assertEquals(s1, interned);
+    //but should be different instance
+    assertNotSame(s1, interned);
+    //asking for it again should again return a different instance
+    final String interned2 = interner.getString(bytes, 0, bytes.length, encoding);
+    assertEquals(s1, interned2);
+    assertNotSame(s1, interned2);
+    assertNotSame(interned, interned2);
+  }
+
+  @Test
+  public void testToString() throws Exception {
+    AsciiStringInterner interner = new AsciiStringInterner();
+    assertEquals("empty", "AsciiStringInterner []", interner.toString());
+    interner.putString("s1");
+    assertEquals("empty", "AsciiStringInterner ['s1']", interner.toString());
+    interner.getString("s2".getBytes(StandardCharsets.US_ASCII), 0, 2, null);
+    assertEquals("empty", "AsciiStringInterner ['s1', 's2']", interner.toString());
+  }
+
+  @Test
+  public void testGarbageCleaning() throws Exception {
```

Comment:
Add `@Category(SlowTests.class)` to this.

Reply:
What is the threshold for slow? This takes 6 seconds for me locally.